src/sources/mod.rs

```diff
@@ -168,24 +169,34 @@ pub trait EventSource {
     /// and [`EventSource::before_handle_events`] notifications. These are opt-in because
     /// they require more expensive checks, and almost all sources will not need these notifications
     const NEEDS_EXTRA_LIFECYCLE_EVENTS: bool = false;
-    /// Notification that a single `poll` is about to begin.
-    /// If this returns Ok(Some), polling will shortcircuit, and
-    /// your event handler will be called with the returned Token and Readiness
+    /// Notification that a single `poll` is about to begin
     ///
     /// Use this to perform operations which must be done before polling,
     /// but which may conflict with other event handlers. For example,
-    /// if polling requires a lock to be taken.
+    /// if polling requires a lock to be taken
+    ///
+    /// If this returns Ok(Some), this will be treated as an event arriving in polling, and
+    /// your event handler will be called with the returned `Token` and `Readiness`.
+    /// Polling will however still occur, so additional events from this or other sources
+    /// may also be handled in the same loop.
+    /// The returned `Token` must belong to this source
+    // If you need to return multiple synthetic events from this notification, please
+    // open an issue
     fn before_sleep(&mut self) -> crate::Result<Option<(Readiness, Token)>> {
         Ok(None)
     }
-    /// Notification that polling is completed, and the resulting events are
-    /// going to be executed.
+    /// Notification that polling is complete, and [`EventSource::process_events`] will
+    /// be called with the given events for this source. The iterator may be empty,
+    /// which indicates that no events were generated for this source
+    ///
+    /// Please note, the iterator will also include any synthetic event returned from
+    /// [`EventSource::before_sleep`]
```

Comment:
Do we actually want that?

The source can already internally track if it gave a synthetic event or not, and at least for the wayland socket, it'd be more useful in this method to know if other *real* events occurred during the polling.

In both cases the wayland source will be able to work with it, but it seems to me including the synthetic events in this iterator only increases the complexity of implementing the sources.